bftengine/src/bftengine/KeyExchangeManager.cpp

```diff
@@ -214,29 +241,31 @@ void KeyExchangeManager::exchangeTlsKeys(const SeqNum& bft_sn) {
   exchangeTlsKeys("server", bft_sn);
   exchangeTlsKeys("client", bft_sn);
   metrics_->tls_key_exchange_requests_++;
-  bft::communication::StateControl::instance().restartComm(repID_);
   LOG_INFO(KEY_EX_LOG, "Replica has generated a new tls keys");
+  bft::communication::StateControl::instance().restartComm(repID_);
+  LOG_INFO(KEY_EX_LOG, "Replica communication restarted after tls exchange");
 }
 void KeyExchangeManager::sendMainPublicKey() {
+  auto [seq, latestPublicKey] = SigManager::instance()->getMyLatestPublicKey();
```

Comment:
If the feature is disabled, don't we need to generate a new main key first?

Reply:
There are no rotations when the feature is disabled, but this is a bug because we this code will crash when the feature is turned off but the publish-main-key-on-startup flag is on. Nice catch.